src/dawn/Optimizer/PassSetBoundaryCondition.cpp

```diff
@@ -0,0 +1,332 @@
+        stencilInstantiation->getOriginalNameFromAccessID(ID));
+    if(it != stencilInstantiation->getNameToAccessIDMap().end()) {
+      if(stencilInstantiation->isField(ID)) {
+        return stencilInstantiation->getAccessIDFromName(
+            stencilInstantiation->getOriginalNameFromAccessID(ID));
+      } else {
+        return (int) FieldType::FT_NotOriginal;
+      }
+    } else {
+      return (int) FieldType::FT_NotOriginal;
+    }
+
+  };
+
+  std::unordered_map<int, Extents> dirtyFields;
+  std::unordered_map<int, std::shared_ptr<BoundaryConditionDeclStmt>> allBCs;
+
+  // Fetch all the boundary conditions stored in the instantiation
+  for(const auto& bc : stencilInstantiation->getBoundaryConditions()) {
+    auto it = stencilInstantiation->getNameToAccessIDMap().find(bc.first);
+    if(it != stencilInstantiation->getNameToAccessIDMap().end()) {
+      int ID = stencilInstantiation->getAccessIDFromName(bc.first);
+      allBCs.emplace(ID, bc.second);
+    }
+  }
+
+  // Get the order in which the stencils are called:
+  VisitStencilCalls findStencilCalls;
+
+  for(const std::shared_ptr<Statement>& statement :
+      stencilInstantiation->getStencilDescStatements()) {
+    statement->ASTStmt->accept(findStencilCalls);
+  }
+  std::vector<int> StencilIDsVisited_;
+  for(const auto& stencilcall : findStencilCalls.getStencilCalls()) {
+    StencilIDsVisited_.push_back(
+        stencilInstantiation->getStencilCallToStencilIDMap().find(stencilcall)->second);
+  }
+
+  auto calculateHaloExtents = [&](std::string fieldname) {
+
+    Extents fullExtent;
+    // Did we already apply a BoundaryCondition for this field?
+    // This is the first time we apply a BC to this field, we traverse all stencils that were
+    // applied before
+    std::vector<int> stencilIDsToVisit(StencilIDsVisited_);
+    if(StencilBCsApplied_.count(fieldname) == 0) {
+    } else {
+      for(int traveresedID : StencilBCsApplied_.find(fieldname)->second) {
+        stencilIDsToVisit.erase(
+            std::remove(stencilIDsToVisit.begin(), stencilIDsToVisit.end(), traveresedID),
+            stencilIDsToVisit.end());
+      }
+    }
+    for(auto& stencil : stencilInstantiation->getStencils()) {
+      for(const int& ID : stencilIDsToVisit) {
+        if(ID == stencil->getStencilID()) {
+          fullExtent.merge(
+              analyzeStencilExtents(stencil, stencilInstantiation->getAccessIDFromName(fieldname)));
+          if(StencilBCsApplied_.count(fieldname) == 0) {
+            StencilBCsApplied_.emplace(fieldname, std::vector<int>{stencil->getStencilID()});
+          } else {
+            StencilBCsApplied_.find(fieldname)->second.push_back(stencil->getStencilID());
+          }
+          break;
+        }
+      }
+    }
+    return fullExtent;
+  };
+
+  auto insertExtentsIntoMap = [](int fieldID, Extents extents, std::unordered_map<int, Extents>& map){
+      auto fieldExtentPair = map.find(fieldID);
```

Comment:
naming convention: the result of a find is not a pair, so I would better name after iterator.
In any case I would recommend count, rather than
```
it = map.find(key)
it != map.end()
``` 
unless `it` is used later on